cpp/src/copying/scatter.cu

```diff
@@ -329,6 +332,104 @@ std::vector<std::unique_ptr<table>> scatter_to_tables(
     input, partition_map, mr, stream);
 }
 
+std::unique_ptr<column> boolean_mask_scatter(
+    column_view const& source, column_view const& target,
+    column_view const& boolean_mask,
+    rmm::mr::device_memory_resource *mr,
+    cudaStream_t stream) {
+
+    auto indices = cudf::make_numeric_column(data_type{INT32},
+                                  target.size(), UNALLOCATED, stream, mr);
+    auto mutable_indices = indices->mutable_view();
+
+    thrust::sequence(rmm::exec_policy(stream)->on(stream),
+                     mutable_indices.begin<size_type>(),
+                     mutable_indices.end<size_type>(),
+                     0);
+
+    // The scatter map is actually a table with only one column, which is scatter map.
+    auto scatter_map = detail::apply_boolean_mask(table_view{{indices->view()}}, boolean_mask,
+                                                  mr, stream);
+    auto output_table =  detail::scatter(table_view{{source}}, scatter_map->get_column(0).view(),
+                                             table_view{{target}}, false, mr, stream);
+
+    // There is only one column in output_table
+    return std::make_unique<column>(std::move(output_table->get_column(0)));
+}
+
+std::unique_ptr<column> boolean_mask_scatter(
+    scalar const& source, column_view const& target,
+    column_view const& boolean_mask,
+    rmm::mr::device_memory_resource *mr,
+    cudaStream_t stream) {
+  
+    return detail::copy_if_else(source, target, boolean_mask, mr, stream);
+}
+
+std::unique_ptr<table> boolean_mask_scatter(
+    table_view const& source, table_view const& target,
+    column_view const& boolean_mask,
+    rmm::mr::device_memory_resource *mr,
+    cudaStream_t stream) {
+
+    CUDF_EXPECTS(source.num_columns() == target.num_columns(), "Mismatch in number of source columns and target columns");
+    CUDF_EXPECTS(boolean_mask.size() == target.num_rows(), "Boolean mask size and number of target rows mismatch");
+    CUDF_EXPECTS(boolean_mask.type().id() == BOOL8, "Mask must be of Boolean type");
+    std::vector<cudf::size_type> column_indices(target.num_columns());
+    std::iota(std::begin(column_indices), std::end(column_indices), 0);
+    // Count valid pair of source and columns as per type at each column index i
+    auto valid_pair_count = std::count_if(column_indices.begin(), column_indices.end(), [&source, &target](auto index){
```

Comment:
I think you can  use a thrust counting_iterator here rather than materializing a vector and using `iota`. It should work with `std` algorithms.

Reply:
Also, instead of `std::count_if()`, use `std::all_of()`
```suggestion
    CUDF_EXPECTS(std::all_of(thrust::counting_iterator<size_type>(0),
                             thrust::counting_iterator<size_type>(target.num_columns()),
                             [&source, &target](auto index){
                               return ((source.column(index).type().id()) == (target.column(index).type().id()));
                             }), 
                  "Type mismatch in source scalar and target column");
```